### DISCUSSION TITLE
Is that pure?

### DISCUSSION THREAD

**Discussion by uniho** [pmndrs/jotai#2769] [General]
https://github.com/pmndrs/jotai/blob/6f7551c0f1fb1b5317ed195882c65bf7caf1850f/src/react/useAtomValue.ts#L136-L140

Is that code pure?
[A React doc](https://react.dev/learn/keeping-components-pure) says "each component should only “think for itself”, and not attempt to coordinate with or depend upon others during rendering.".
That code may be on "during rendering".

**Comment by dai-shi** [comment]
It's pure or idempotent.
Setting local state during rendering is allowed in the docs.
However, this doesn't solve tearing. For that, useSyncExternalStore would be necessary.

**Reply by uniho** [reply]
@dai-shi 

Thanks.
If it's pure then it's fine.
I realize that the reason you don't use use**Sync**ExternalStore() is because use**Sync**ExternalStore() doesn't correspond to `startTransition` **async**.
